syntax = "proto3";

package api.record;

//import "google/protobuf/empty.proto";
//import "google/protobuf/timestamp.proto";
import "google/protobuf/field_mask.proto";

service Record {
  rpc GetById (RecordsByIdRequest) returns (RecordDataList);

  rpc GetWithPagination(RecordsPaginationRequest) returns (RecordsCursorResponse);

  rpc Insert (EditRecordsRequest) returns (RecordDataList);
  rpc Update (EditRecordsRequest) returns (StatusResponse);

  rpc RemoveById (RecordsByIdRequest) returns (StatusResponse);
}

message RecordData {
  string id = 1;
  string createdAt = 2;
  string lastUpdate = 3;
  string profileId = 4;
  string record = 5;
}

message RecordDataList {
  repeated RecordData records = 1;
}

//---

message StatusResponse {
  string message = 1;
}

message RecordsCursorResponse {
  RecordDataList recordList = 1;
  string cursor = 2;
}

//---

message RecordsByIdRequest {
  repeated string ids = 1;
  google.protobuf.FieldMask fields = 2;
}

message EditRecordsRequest {
  RecordDataList recordList = 1;
  google.protobuf.FieldMask fields = 2;
}

enum RecordKey {
  UNKNOWN = 0;
  ID = 1;
  CREATEDAT = 2;
  LASTUPDATE = 3;
  PROFILEID = 4;
  RECORD = 5;
}

enum Operand {
  NONE = 0;
  IN = 1; // in or ==
  OUT = 2; // out or !=
  LESS = 3; // less or <
  GREATER = 4; // greater or >
  EQLESS= 5; // not greater or <=
  EQGREATER = 6; // not less or >=
}

enum Direction {
  NOT = 0;
  MIN = 1;
  MAX = 2;
}


message Filter {
  RecordKey key = 1;
  Operand operand = 2;
  repeated string values = 3;
}

message Ordering {
  RecordKey key = 1;
  Direction direction = 2; 
}

message RecordsPaginationRequest {
  string cursor = 1;

  repeated Filter filters = 2;
  repeated Ordering orders = 3;

  int32 limit = 4;
}